worlds/tww/__init__.py

```diff
@@ -0,0 +1,577 @@
+
+        # Output the plando details to file.
+        aptww = TWWContainer(
+            path=os.path.join(
+                output_directory, f"{multiworld.get_out_file_name_base(player)}{TWWContainer.patch_file_ending}"
+            ),
+            player=player,
+            player_name=self.player_name,
+            data=output_data,
+        )
+        aptww.write()
+
+    def extend_hint_information(self, hint_data: dict[int, dict[int, str]]) -> None:
+        """
+        Fill in additional information text into locations, displayed when hinted.
+
+        :param hint_data: A dictionary of mapping a player ID to a dictionary mapping location IDs to the extra hint
+        information text. This dictionary should be modified as a side-effect of this method.
+        """
+        # Create a mapping of island names to numbers for sunken treasure hints.
+        island_name_to_number = {v: k for k, v in ISLAND_NUMBER_TO_NAME.items()}
+
+        hint_data[self.player] = {}
+        for location in self.multiworld.get_locations(self.player):
+            if location.address is not None and location.item is not None:
+                # Regardless of ER settings, always hint at the outermost entrance for every "interior" location.
+                zone_exit = self.entrances.get_zone_exit_for_item_location(location.name)
+                if zone_exit is not None:
+                    outermost_entrance = self.entrances.get_outermost_entrance_for_exit(zone_exit)
+                    assert outermost_entrance is not None and outermost_entrance.island_name is not None
+                    hint_data[self.player][location.address] = outermost_entrance.island_name
+
+                # Hint at which chart leads to the sunken treasure for these locations.
+                if location.name.endswith(" - Sunken Treasure"):
+                    island_name = location.name.removesuffix(" - Sunken Treasure")
+                    island_number = island_name_to_number[island_name]
+                    chart_name = self.charts.island_number_to_chart_name[island_number]
+                    hint_data[self.player][location.address] = chart_name
+
+    def determine_item_classification(self, name: str) -> IC | None:
+        """
+        Determine the adjusted classification of an item. The classification of an item may be affected by which options
+        are enabled or disabled.
+
+        :param name: The name of the item.
+        :return: The adjusted classification of the item. If there is no adjustment from the default, returns `None`.
+        """
+        # TODO: Calculate nonprogress items dynamically
+        adjusted_classification = None
+        if not self.options.progression_big_octos_gunboats and name == "Quiver Capacity Upgrade":
+            adjusted_classification = IC.useful
+        if self.options.sword_mode == "swords_optional" and name == "Progressive Sword":
+            adjusted_classification = IC.useful
```

Comment:
Should `"Progressive Sword"` also be changed to `useful` in swordless mode?

If they were, then the `has_heros_sword()`, `has_any_master_sword()` and `has_full_power_master_sword()` macros would not need to check `not state._tww_in_swordless_mode(player)` because only items with the `progression` classification (or a mix of classifications that includes `progression`) are considered by logic.

Reply:
This may not be necessary since swords wouldn't be placed in swordless mode, but it doesn't hurt.

Reply:
That's fair. I hadn't considered those cases.